## Build a Delft3D FM 1D2D model from scratch

This notebook demonstrates how to prepare a **1D2D Delft3D FM** model from scratch using the command line interace (CLI). 

We will combine elements that were prepared in the *Build 1D model* and *Build 2D model* notebooks, so we advise you to have a look at these before starting this one.

All lines in this notebook which start with `!` are executed from the command line. Within the notebook environment the logging messages are shown after completion. You can also copy these lines and paste these in your shell to get more direct feedback.

### HydroMT CLI build interface

Lets first check if the Delft3D FM model is recognized by HydroMT:

In [1]:
# this should return "dflowfm" to build DFlowFM components of Delft3D FM
# as well as the generic HydroMT models "grid_model, lumped_model, mesh_model, network_model"
!hydromt --models

model plugins:
 - dflowfm (hydromt_delft3dfm 0.1.0)
generic models (hydromt 0.8.1.dev0):
 - grid_model
 - lumped_model
 - mesh_model
 - network_model



Using the **HydroMT build** API we can setup a complete model from scratch. Let's get an overview of all the available options:

In [2]:
!hydromt build --help

Usage: hydromt build [OPTIONS] MODEL MODEL_ROOT

  Build models from scratch.

  Example usage: --------------

  To build a wflow model for a subbasin using a point coordinates snapped to
  cells with upstream area >= 50 km2 hydromt build wflow /path/to/model_root
  -i /path/to/wflow_config.ini  -r "{'subbasin': [-7.24, 62.09], 'uparea':
  50}" -d deltares_data -d /path/to/data_catalog.yml -v

  To build a sfincs model based on a bbox hydromt build sfincs
  /path/to/model_root  -i /path/to/sfincs_config.ini  -r "{'bbox':
  [4.6891,52.9750,4.9576,53.1994]}"  -d /path/to/data_catalog.yml -v

Options:
  --opt TEXT               Method specific keyword arguments, see the method
                           documentation of the specific model for more
                           information about the arguments.
  -i, --config PATH        Path to hydroMT configuration file, for the model
                           specific implementation.
  -r, --region TEXT        Set the region for which to 

### Data for model setup

Most of the time, the `<...>_fn` arguments correspond to a data source from the `DataCatalog` which is based on a yml-file with references to the data paths/urls and how the data should be read. This file can be provided to hydromt using `-d <path_to_yml_file>`. 

By default, the pre-defined `artifact_data` catalog containing some example data for the Piave basin will be downloaded to `~/.hydromt_data/` which is also used for this example. An overview of the available example data is provided [here](https://deltares.github.io/hydromt/latest/user_guide/data_existing_cat.html)
This example data is a based on the data which is available from the Deltares p-drive. 
If you have acces to this drive, a pre-configured catalog file can be loaded using the `-d deltares_data` pre-defined catalog. 

More information on how to write a data catalog yml file can be found in the [HydroMT core docs](https://deltares.github.io/hydromt/latest/user_guide/data_prepare_cat.html) 

> **NOTE**: In HydroMT-Delft3D FM, an additionnal data catalog with all default parameters values is always used [parameters_data.yml](https://github.com/Deltares/hydromt_delft3dfm/blob/54-examples/hydromt_delft3dfm/data/parameters_data.yml). It contains default values for 1D network elements and structures as well as default 2D landuse paramter mapping tables.

### Model setup configuration

The HydroMT configuration file contains the model setup configuration and determines which methods are used to prepare the different components of a Delft3D FM model and in which order and optionally sets non-default arguments for each method. This configuration is passed to HydroMT using `-i <path_to_configuration_file>`. The supported format are either YAML or INI. We have prepared several example yml-files which are available in the model repository [examples folder](https://github.com/Deltares/hydromt_delft3dfm/tree/main/examples) and from the [docs (building a model)](https://deltares.github.io/hydromt_delft3dfm/latest/user_guide/dflowfm_build). 

Each section, before indent, (e.g. setup_rivers_from_dem) corresponds to a model method. All model methods are explained in the [docs (model components)](https://deltares.github.io/hydromt_delft3dfm/latest/user_guide/dflowfm_model_setup.html). The `global` section contains direct model initialisation properties.

We will load the default dflowfm build 1D2D yaml file for inspection:

In [3]:
fn_yml = "dflowfm_build.yml"
with open(fn_yml, "r") as f:
    txt = f.read()
print(txt)

global:
  crs: 3857
  network_snap_offset: 25
  openwater_computation_node_distance: 40

setup_rivers_from_dem:
  region: 
    bbox: [12.4331, 46.4661, 12.5212, 46.5369]
  hydrography_fn: merit_hydro
  river_geom_fn: rivers_lin2019_v1
  rivers_defaults_fn: rivers_defaults
  rivdph_method: gvf
  rivwth_method: geom
  river_upa: 25.0
  friction_type: "Manning"
  friction_value: 0.023
  rivbankq: 25

setup_pipes:
  region: 
    bbox: [12.4331, 46.4661, 12.5212, 46.5369]
  pipes_fn: grip_roads
  pipes_defaults_fn: pipes_defaults
  pipe_filter: pipe
  spacing: 50
  friction_type: WhiteColeBrook
  friction_value: 0.003
  crosssections_shape: circle
  crosssections_value: 0.5
  dem_fn: merit_hydro # [copdem30, merit]
  pipes_depth: 2.0
  snap_offset: 0.5
  pipes_invlev: 3. 

setup_manholes:
  manholes_fn:
  manhole_defaults_fn: manholes_defaults
  dem_fn: merit_hydro
  bedlevel_shift: 0.5

setup_1dboundary:
  boundary_value: -2.0
  branch_type: river
  boundary_type: waterlevel
  boundary_uni

Looking at this file, you see that we will prepare the following information for our model:

- [global](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.html#hydromt_delft3dfm.DFlowFMModel): our model will be defined in the projected CRS WGS84 EPSG 3857.
- [setup_rivers_from_dem](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_rivers_from_dem.html): we will derive the 1D rivers lines based on the *MERIT Hydro DEM* and that intersects with the region bounding box [12.4331, 46.4661, 12.5212, 46.5369]. River roughness will be a Manning constant of 25.0 and we will use data from *rivers_lin2019* database to prepare rectangular cross-sections. Other default parameters are read from rivers_defaults data.
- [setup_pipes](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_pipes.html): we will add pipes by assuming that they are located under the roads as defined in the *grip_roads* data. Roughness will be a WhiteColeBrook constant of 0.003 and we will use a circular cross-sections of 0.5m. The pipes invert levels will be derived from the *MERIT Hydro DEM* assuming a constant depth of the pipe underground of 2 meters. Other default parameters are read from pipes_defaults data.
- [setup_manholes](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_manholes.html): for a dflowfm model to be valid, if pipes are defined, then the corresponding should be added as well (careful of the order, setup_manholes can only be called after all elements of the network have been prepared, ie setup_pipes, setup_rivers and/or setup_rivers_from_dem and/or setup_channels). Here we do not have specific input data locations for our pipes, so manholes will be generated based on a set of standards specified in manhole_defaults_fn.
- [setup_1dboundary](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_1dboundary.html): finally as a default, we will initialise the upstream and downstream boundaries of our rivers with a constant waterlevel of -2.0 meter a.s.l
- [setup_mesh2d](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_mesh2d.html): we will prepare a regular grid for the region bounding box [12.4331, 46.4661, 12.5212, 46.5369] of 500*500 meters resolution.
- [setup_maps_from_rasterdataset](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_maps_from_rasterdataset.html): prepares gridded elevation from *merit_hydro*. Here elevation is only extracted from *merit_hydro* and later the Delft3D FM kernel will interpolate it to the mesh 2D grid using the *nearestNb* method.
- [setup_maps_from_raster_reclass](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_maps_from_raster_reclass.html): prepares gridded Manning roughness and infiltration capacity data by reclassifying the landuse categories in *vito* landuse data, using parameters values defined per landuse category in *vito_mapping* table. When running, the Delft3D FM kernel will interpolate it to the mesh 2D grid using the *triangulation* method.

Additionnally to a 1D or 2D models only, we have now added a new method:

- [setup_link1d2d](https://deltares.github.io/hydromt_delft3dfm/latest/_generated/hydromt_delft3dfm.DFlowFMModel.setup_link1d2d.html): we will link our 1D and 2D mesh by setting up links from the 1D network to the 2D grid. As you can imagine, this function can only be run once both the schematisation of the 1D and the 2D grids are finalized. If any change is made to the 1D or the 2D grid, this funcion should be run again.

Feel free to use the links of each components to know more about how each function works and what are all the available settings.

> **NOTE**: 
  In HydroMT Delft3D FM, there is not a unique region definition. For example, the extent of the 1D river can differ from the extent of the 1D urban network or the 2D inundation grid. For this reason, the user should specify for each of the setup method defining the 1D or 2D grid definition the corresponding region. Also the user should specify the dflowfm model destination CRS in the `global` section in order to allow to use grid definition data (rivers, pipes) from different data sources or CRS. 

### Building a 1D2D Delft3D FM model

In [4]:
# NOTE: copy this line (without !) to your shell for more direct feedback - should take about 1 minute or 2 to run
!hydromt build dflowfm "./build/dflowfm_1d2d" -i dflowfm_build.yml -d artifact_data --fo -vv

2023-09-22 12:33:33,963 - build - log - DEBUG - Writing log messages to new file C:\Users\boisgont\D\Repos\hydromt\hydromt_delft3dfm\examples\build\dflowfm_1d2d\hydromt.log.
2023-09-22 12:33:33,963 - build - log - INFO - HydroMT version: 0.8.1.dev0
2023-09-22 12:33:33,963 - build - main - INFO - Building instance of dflowfm model at C:\Users\boisgont\D\Repos\hydromt\hydromt_delft3dfm\examples\build\dflowfm_1d2d.
2023-09-22 12:33:33,963 - build - main - INFO - User settings:
2023-09-22 12:33:37,232 - build - data_catalog - INFO - Reading data catalog artifact_data v0.0.8 from archive
2023-09-22 12:33:37,232 - build - data_catalog - INFO - Parsing data catalog from C:\Users\boisgont\.hydromt_data\artifact_data\v0.0.8\data_catalog.yml
2023-09-22 12:33:37,399 - build - model_api - INFO - Initializing dflowfm model from hydromt_delft3dfm (v0.1.0).
2023-09-22 12:33:37,399 - build - data_catalog - INFO - Parsing data catalog from C:\Users\boisgont\D\Repos\hydromt\hydromt_delft3dfm\hydromt_del

The nodata value None is not in the reclass table.None will be used for the params.


The example above means the following: run **hydromt build** with:

* `dflowfm` : i.e. build a FM model
* `./build/dflowfm_1d2d` : output model folder
* `-i dflowfm_build.yml`: model build configuration file
* `-d artifact_data`: data catalog to use. Here the default artifact data catalog with global data for Northern Italy.
* `--fo`: force overwritting in case a model already exists in the output folder.
* `-vv` : give some extra verbosity (2 * v) to display feedback on screen. Now debug and info messages are provided rather than only warnings and errors.

> NOTE: We did not specify a region `-r` as with dflowfm, the region is specific to each of the 1D network elements we would like to add, in this example rivers and pipes.

> NOTE: You can use as many data catalogs as you need within the same command line, by repeating the `-d` options and name of the catalog to use several times.

Next we check which files have been created. The model root should contain three files and three folders:

- dimr_config.xml: DIMR file to execute Delft3D FM from command line.
- hydromt.log and hydromt_data.yml: information saved by HydroMT on model build settings and used datasets for reproducibility
- **dflowfm** folder: dflowfm files, here for our 1D model. Contains all necessary information to run a 1D2D model including the mesh 1D 2D and links (fm_net.nc file), roughness, cross-sections, 1D boundary forcing as well as the settings and paths to the mesh2d data files in the maps folder.  The configuration MDU file is also created as well as a *.gui* file to visualize the 1D network in the Delft3D FM GUI.
- **geoms** folder: folder containing a copy of the dflowfm data in an easily readable GIS format rather than the ini or UGRID netcdf file.
- **maps** folder: folder containing regular gridded data that Delft3D FM can interpolate to the 2D mesh grid. Here we see our elevation or bedlevel data, as well as the roughness / friction and infiltration capacity maps.

In [5]:
import os

root = "./build/dflowfm_1d2d/"
for path, _, files in os.walk(root):
    print(path)
    for name in files:
        print(f" - {name}")

./build/dflowfm_1d2d/
 - dimr_config.xml
 - hydromt.log
 - hydromt_data.yml
./build/dflowfm_1d2d/dflowfm
 - bnd.ext
 - boundaryconditions1d_waterlevelbnd.bc
 - branches.gui
 - crsdef.ini
 - crsloc.ini
 - DFlowFM.mdu
 - fieldFile.ini
 - fm_net.nc
 - nodeFile.ini
 - roughness_M-0p023.ini
 - roughness_W-0p003.ini
./build/dflowfm_1d2d/geoms
 - boundaries.geojson
 - branches.geojson
 - crosssections.geojson
 - manholes.geojson
 - mesh1d.geojson
 - mesh2d.geojson
 - network1d.geojson
 - pipes.geojson
 - pipe_nodes.geojson
 - region.geojson
 - rivers.geojson
 - rivers_nodes.geojson
./build/dflowfm_1d2d/maps
 - bedlevel.tif
 - frictioncoefficient.tif
 - InfiltrationCapacity.tif


### Visualize and/or inspect model schematization

* The **dflowfm plot** example notebook contains scripts to visualize your model.